staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go

```diff
@@ -183,73 +183,141 @@ func TestListenerResyncPeriods(t *testing.T) {
+	if e, a := zeroDuration(5*time.Second, !doResyncs), informer.resyncCheckPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := time.Duration(0), informer.processor.listeners[0].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := 1*time.Minute, informer.processor.listeners[1].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := 55*time.Second, informer.processor.listeners[2].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	if e, a := 5*time.Second, informer.processor.listeners[3].requestedResyncPeriod; e != a {
+		t.Errorf("expected %d, got %d", e, a)
+	}
+	checkResyncPeriods(t, informer.processor, !doResyncs)
+
+	// Now start the informer so we can test that the resyncCheckPeriod shrinks no further
+	stop := make(chan struct{})
+	defer close(stop)
+	go informer.Run(stop)
+
+	// Now wait for the informer to finish starting up
+	if !listener1.ok() {
+		t.Errorf("%s: expected %v, got %v", listener1.name, listener1.expectedItemNames, listener1.receivedItemNames)
+	}
+
+	// listener 5, request resync every 3 seconds, get every 5
+	listener5 := newTestListener("listener5", 3*time.Second, "pod1")
+	informer.AddEventHandlerWithResyncPeriod(listener5, listener5.resyncPeriod)
+	if e, a := zeroDuration(5*time.Second, !doResyncs), informer.resyncCheckPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := time.Duration(0), informer.processor.listeners[0].resyncPeriod; e != a {
+	if e, a := time.Duration(0), informer.processor.listeners[0].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := 1*time.Minute, informer.processor.listeners[1].resyncPeriod; e != a {
+	if e, a := 1*time.Minute, informer.processor.listeners[1].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := 55*time.Second, informer.processor.listeners[2].resyncPeriod; e != a {
+	if e, a := 55*time.Second, informer.processor.listeners[2].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
-	if e, a := 5*time.Second, informer.processor.listeners[3].resyncPeriod; e != a {
+	if e, a := 5*time.Second, informer.processor.listeners[3].requestedResyncPeriod; e != a {
 		t.Errorf("expected %d, got %d", e, a)
 	}
+	if e, a := ifElseDuration(doResyncs, 5*time.Second, 3*time.Second), informer.processor.listeners[4].requestedResyncPeriod; e != a {
+		t.Errorf("for doResyncs=%v: expected %d, got %d", doResyncs, e, a)
+	}
+	checkResyncPeriods(t, informer.processor, !doResyncs)
+
+}
+
+func checkResyncPeriods(t *testing.T, p *sharedProcessor, expectZero bool) {
+	for idx, listener := range p.listeners {
+		e := zeroDuration(listener.requestedResyncPeriod, expectZero)
+		a := listener.resyncPeriod
+		if e != a {
+			t.Errorf("for listener %d: expected %d, got %d", idx, e, a)
+		}
+	}
+}
+
+func zeroDuration(dur time.Duration, zeroIt bool) time.Duration {
```

Comment:
what is this trying to do?  The fact that I have to ask probably isn't good.  I suggesting renaming or refactoring.

Reply:
I renamed and added an explanatory comment.